libraries/lib-wave-track/WaveClip.cpp

```diff
@@ -1140,37 +1140,56 @@ void WaveClip::ApplyStretchRatio(const ProgressReporter& reportProgress)
    auto newSequences = GetEmptySequenceCopies();
    bool swappedOnce = false;
 
-   Finally Do { [&, trimLeftBeforeStretch = mTrimLeft,
-                 trimRightBeforeStretch = mTrimRight] {
-      // Whether successful or not, the right thing to do is to restore the
-      // original trim values.
-      this->SetTrimLeft(trimLeftBeforeStretch);
-      this->SetTrimRight(trimRightBeforeStretch);
+   Finally Do { [&, trimLeftBeforeStretch = GetTrimLeft(),
+                 trimRightBeforeStretch = GetTrimRight(),
+                 offsetBeforeStretch = GetSequenceStartTime()] {
       if (success)
       {
          this->mClipStretchRatio = 1.0;
          this->mRawAudioTempo = this->mProjectTempo;
          assert(this->GetStretchRatio() == 1.0);
       }
-      else if (swappedOnce)
-         std::swap(mSequences, newSequences);
+      else
+      {
+         this->SetTrimLeft(trimLeftBeforeStretch);
+         this->SetTrimRight(trimRightBeforeStretch);
+         this->SetSequenceStartTime(offsetBeforeStretch);
+         if (swappedOnce)
+            std::swap(mSequences, newSequences);
+      }
    } };
 
-   SetTrimLeft(0);
-   SetTrimRight(0);
+   const auto originalPlayStartTime = GetPlayStartTime();
+   const auto originalPlayEndTime = GetPlayEndTime();
+
+   // Leave 1 second of raw, unstretched audio before and after visible region
+   // to ensure the algorithm is in a steady state when reaching the play
+   // boundaries.
+   TrimLeftTo(originalPlayStartTime - stretchRatio);
+   TrimRightTo(originalPlayEndTime + stretchRatio);
```

Comment:
Is this the smartest way to leave sufficient time to find the steady state?  Deeper knowledge of the library might be needed to answer that.

Should you reserve a certain amount of time or calculcate how to allow a certain minimum number of analysis windows?

Do you really need at the right end to?  Is the algorithm influenced by some lookahead?  (As with a compressor)

Reply:
It's a rather quick-and-dirty approach, given. It's probably an overkill, and 1 complete FFT size worth of samples should be sufficient. But we may get the FFT size to become sample-rate dependent in the future (there's an issue for that) and a self-maintaining solution would mean communicating that value somehow. Except for extreme stretch ratios, I don't expect this overkill to have a noticeable impact on performance.